### PR TITLE
Morphs and Partially draw Objects

### PR DESCRIPTION
**Pull request by ArbitRandomUser**
- [x] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [x] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [x] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [x] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [x] Did I properly add test dependencies to the `test` directory (if applicable)?
- [x] Did I check relevant tutorials that may be affected by changes in this PR?
- [x] Did I clearly articulate why this PR was made the way it was and how it was made?

**How did you address these issues with this PR? What methods did you use?**

> ### Install master branch [Luxor.jl](https://github.com/JuliaGraphics/Luxor.jl) branch till Luxor 3.5 is released if you want to try out this PR

PR allows Arbitrary morphs from any `Object` to `Object` , `Object` to `Function` . 
and also keyframed  morphs with `Animations.jl` .
PR also brings in ability to draw Objects partially , and "show-creation" (and destruction) of objects.

This is made possible with the help of JPath a structure holding data of polygons and how to fill and stroke them.
Every Javis `Object` now has a a `jpaths` field an array for `JPath` that is got from evaluating the `object.func` fo the `Object`

Extracting the polygons is made possible by over-riding luxors strokepath/stroke/fillpath/fillpreserve methods . These four methods are only ways to lay the current path onto the canvas and all of luxors "drawing function" eventually call one or more of these four. (except for text ).

Morphs between polygons are using a modified version of luxors polymorph. (modified so that
polys dont have to be resampled everytime its called)

**Limitations:**
afaik...

1) Cannot be used in conjunction with `sethue` in Actions. Morphs will work , but the sethue will not
if you want to change hue while morphing you will have to change the hue of the target you are morphing to.
2) Objects and Functions which have clipping regions in their drawings will not be morphed properly.
3) Gets the morphs done .. ~~but might not be the most "natural" or simplest morph possible.~~ morphs are somewhat "natural" now by offseting the index for best match between two polygons. But individual polygons are not matched for the shortest transformation
4) dashed lines not yet implemented . Objects should use only solid lines
5) strokescale doesnt change during morph (not yet implemented) . Objects should have the default strokethickness 

### `Object` to `Object` morphs ...

```julia
using Javis

video = Video(500,500)
nframes = 160 

function circdraw(colo)
    sethue(colo)
    setopacity(0.5)
    circle(O,100,:fillpreserve)
    setopacity(1.0)
    sethue("white")
    strokepath()
end

function boxdraw(colo)
    sethue(colo)
    box(O,100,100,:fillpreserve)
    setopacity(1.0)
    sethue("white")
    strokepath()
end

Background(1:nframes,(args...)->background("black"))
boxobj  = Object((v,o,f) -> boxdraw("green"))
circobj = Object((v,o,f) -> circdraw("red"))

transform_to_box = Action(20:nframes-20, morph_to(boxobj))
act!(circobj, transform_to_box)
render(video,pathname="circ_to_box.gif")


```

circ_to_box_pr

### `Object` to `Function` morphs ...

```julia
Background(1:nframes,(args...)->background("black"))
circobj = Object(1:nframes,(args...) -> circdraw("red"))

transform_to_box = Action(20:nframes-20, morph_to(boxdraw,["blue"]))
act!(circobj, transform_to_box)

render(video,pathname="circ_to_box_fn.gif")
```

circ_to_box_fn

### Keyframed morphs using Animations.jl

```julia
using Javis
using Animations
video = Video(500,500)
nframes = 160 

function circdraw(colo)
    sethue(colo)
    setopacity(0.5)
    circle(O,50,:fillpreserve)
    setopacity(1.0)
    sethue("white")
    strokepath()
end

function boxdraw(colo)
    sethue(colo)
    box(O,100,100,:fillpreserve)
    setopacity(1.0)
    sethue("white")
    strokepath()
end

function stardraw()
    sethue("white")
    star(O,100,5,0.5,0.0,:stroke)
end

Background(1:nframes+10,(args...)->background("black"))
boxobj = Object(1:nframes+10 , (args...) -> boxdraw("green") )
anim = Animation([0, 0.7, 1],  MorphFunction[(boxdraw, ["green"]),
                                             (stardraw, []),
                                             (circdraw, ["red"])
                                             ])


action = Action(1:nframes,anim,morph())
act!(boxobj,action)
render(video,pathname="box_to_star_to_circ.gif")
```

box_to_star_to_circ

## Examples for Partial Draw .
```julia
using Javis
video = Video(500,500)
nframes = 360 

function ground()
    background("black")
    sethue("white")
end

function boxdraw(colo)
    circle(O+100,50,:stroke)
    sethue(colo)
    box(O,100,100,:fillpreserve)
    setopacity(1.0)
    sethue("white")
    strokepath()
end

Background(1:nframes+60,(args...)->ground())
boxobj = Object(1:nframes+60 , (args...) -> boxdraw("green") )

action_create = Action(1:nframes÷2,sineio(),Javis.showcreation())
action_remove = Action(nframes÷2+30:nframes,sineio(),Javis.showdestruction())
act!(boxobj,action_create)
act!(boxobj,action_remove)
render(video,tempdirectory="images",pathname="show_box_creation.gif")
run(`mpv show_box_creation.gif`)
```
show_box_creation

### CODE REVIEW

File: src/Javis.jl [hunk 1]
```diff
@@ -561,6 +563,17 @@ Returens the final rendered frame
 """
 function get_javis_frame(video, objects, frame; layers = Layer[])
 
+    #extract object jpaths on first frame
+    if frame == 1
+        for object in objects
+            try
+                println("EXTRACTING " * object.opts[:name] * " JPATHS")
```
Comment (Wikunia) [inline src/Javis.jl:570]: I'm confused by the `try-catch` block here.

File: src/action_animations.jl [hunk 2]
```diff
@@ -486,3 +486,42 @@ function _change(video, object, action, rel_frame, s)
     val = get_interpolation(action, rel_frame)
     object.change_keywords[s] = val
 end
+
+"""
+morph() to be used with Action, when an animation form Animations.jl is
```
Comment (Wikunia) [inline src/action_animations.jl:491]: Before merge: proper docstring with the function definition indented is needed

File: src/luxor_overrides.jl [hunk 3]
```diff
@@ -273,3 +274,298 @@ function background(background_color)
     end
     Luxor.background(background_color)
 end
+
+function apply_transform(transform::Vector{Float64}, poly::Vector{Point})
```
Comment (Wikunia) [inline src/luxor_overrides.jl:278]: We aren't strict with this atm but I would like to avoid having methods in this file that are not actually overrides. Maybe we can have a new file like `luxor_overrides_utils.jl` or something like that.

File: src/structs/JPath.jl [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+- `polys` a list of polygons that makes up the path
+- `closed` a list of bools of same length as `polys`. closed[i] 
+states if polys[i] is a closed polygon or not
+- `fill` a vector of 4 numbers , R, G ,B and A the color it
+should be filled with. if the path was not filled its A is set to 0
+(along with R,G,B) this way its an "invisible" fill.
+- `stroke a vector of 4 numbers just like `fill` for the stroke color.
+- linewidth for stroke 
+"""
+mutable struct JPath
```
Comment (Wikunia) [inline src/structs/JPath.jl:19]: I'm quite confused by the naming here. Every objects can have several `JPaths` but each `JPath` is also a vector of polys? For me a `JPath` sounds like a single path or polygon if closed.

File: src/structs/Object.jl [hunk 5]
```diff
@@ -23,8 +23,10 @@ mutable struct Object <: AbstractObject
     opts::Dict{Symbol,Any}
     change_keywords::Dict{Symbol,Any}
     result::Vector
+    jpaths::Any
```
Comment (Wikunia) [inline src/structs/Object.jl:26]: Is there a reason for `Any` here?

File: src/luxor_overrides.jl [hunk 6]
```diff
@@ -273,3 +274,298 @@ function background(background_color)
...
+    retpoly = Point[]
+    for pt in poly
+        res = cairotojuliamatrix(transform) * [pt[1], pt[2], 1]
+        push!(retpoly, Point(res[1], res[2]))
+    end
+    retpoly
+end
+
+function Luxor.strokepath()
```
Comment (Wikunia) [inline src/luxor_overrides.jl:287]: Feels like the following four functions have a lot of the same function calls. Would be easier to read if another function is added which takes in inputs like `:stroke` or `:fill`

**Review by Wikunia — COMMENTED** [review]
Just some general comments for now @ArbitRandomUser 
Don't worry about them too much now it's just something to keep in mind.
Do you mind adding one or two examples of the new functionality to the PR description to make it easy to test things out? Also some sentences for a description of how it works would be nice to have everything in one place.

File: src/structs/JPath.jl (hunk 4 above)
Comment (ArbitRandomUser) [reply src/structs/JPath.jl:19]: yes .
when a drawing function `obj.func` is called 
every call to strokepath / fillpath will add a new jpath.
so if an object.func does something like ..
`circle(O,100,:stroke)
line(O,100,:stroke)`
it will have 2 jpaths , one for the circle and one for the line.

lets say i instead had stored them as polygons . and i made some object like this .. a ring.
`circlepath(O,100,:path)
cirlcepath(O,50,:path,reversepath=true)
fillpath()
`
if i store every polygon and its associated stroke and fill , and i replay by placing every polygon and 
then immediatly stroke/fill i will not get a ring like the above code would draw.

Effectively an array JPath keeps track of where to call fill/stroke while drawing out the polygons inside it

Comment (Wikunia) [reply src/structs/JPath.jl:19]: Thanks for the explanation. I think it makes more sense to me now. Would it make sense to call it `JPaths` though? As it holds information about several paths?

File: src/Javis.jl (hunk 1 above)
Comment (ArbitRandomUser) [reply src/Javis.jl:570]: this will be removed , just a hack around to  debug  code.

File: src/action_animations.jl (hunk 2 above)
Comment (ArbitRandomUser) [reply src/action_animations.jl:491]: yes , will do

File: src/luxor_overrides.jl (hunk 6 above)
Comment (ArbitRandomUser) [reply src/luxor_overrides.jl:287]: yes a lot of repetitive code for the 4 functions , ill wrap them up to make it cleaner

File: src/structs/JPath.jl (hunk 4 above)
Comment (ArbitRandomUser) [reply src/structs/JPath.jl:19]: maybe we should rename it alltogether.
i would prefer datastructures to have a "singular" name .
this allows little more clarity in name arrays of that structure.

for ex if a structure is called X
.. so arrays of it can be  Xs

but if its already called Xs , arrays of it can be a little confusing

File: src/luxor_overrides.jl [hunk 7]
```diff
@@ -273,3 +274,253 @@ function background(background_color)
...
+end
+
+"""
+updates CURRENT_JPATHS
+if used inside strokepath() pass action=:stroke
+and inside fillpath pass :fill.
+move this to luxor_overrides_util.jl later.
+"""
+function update_currentjpath(action::Symbol)
+    if CURRENT_FETCHPATH_STATE == true
```
Comment (gpucce) [inline src/luxor_overrides.jl:302]: Hi @ArbitRandomUser I think you wrote smth like this a couple of time shouldn't in general
`if a == true` be the same as `if a` and `if a == false` the same as `if !a`

File: src/luxor_overrides.jl [hunk 8]
```diff
@@ -273,3 +274,253 @@ function background(background_color)
...
+        #println("test strokepaths")
```
Comment (gpucce) [inline src/luxor_overrides.jl:303]: Hi @ArbitRandomUser can't it be 
`if CURRENT_FETCHPATH_STATE == true` 
be changed to
`if CURRENT_FETCHPATH_STATE`? 
And similarly `if CURRENT_FETCHPATH_STATE == false` become `if !CURRENT_FETCHPATH_STATE` this is only if they can only be boolen but for 
CURRENT_FETCHPATH_STATE it should be the case?

Comment (ArbitRandomUser) [reply src/luxor_overrides.jl:303]: yes , i will change that .

Comment (ArbitRandomUser) [reply src/luxor_overrides.jl:303]: done , i pushed the changes , dont know why it doesn't show "outdated" above the review comment here

File: src/morphs.jl [hunk 9]
```diff
@@ -42,6 +42,158 @@ function morph_to(to_func::Function; do_action = :stroke)
...
+        end
+    end
+
+    for (jpath1, jpath2) in zip(jpaths1, jpaths2)
+        push!(interp_jpaths, _morph_jpath(jpath1, jpath2, get_interpolation(action, frame)))
+    end
+    function drawfunc(args...)
+        drawjpaths(interp_jpaths)
+        global DISABLE_LUXOR_DRAW = true
+        ret = object.opts[:original_func]()
```
Comment (Wikunia) [inline src/morphs.jl:106]: It's very confusing to have this function in the middle of another function.
It doesn't seem to work like this for your first example as well. Shouldn't it be  `object.opts[:original_func](args...)`?

Comment (Wikunia) [reply src/morphs.jl:106]: Additionally the code you have in the PR description will not produce the gif there, right? Or why should the green square not be there the whole time?

File: docs/src/tutorials/tutorial_morphing.md [hunk 10]
```diff
@@ -0,0 +1,188 @@
+# **Tutorial 9:** Morphing `Javis` Objects   
+
+There are multiple ways to morph an object in Javis.
+
+- Using the older `morph_to(::Function)` method. Very nice at matching the shapes for morphing . Has limitations on what the morphing function should contain.
+- Using (New) `morph_to(::Object)` method. Any Object can be morphed to any other object using this method.
+- Using `morph_to_fn(::Function)` method. Similar to `morph_to(::Object)` but morphs to function instead. Can morph an object to a function that contains Luxor calls to draw what it should morphed into.
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:7]: I think it's quite confusing why we have `morph_to(::Function)` and `morph_to_fn`. Can't they be combined?

File: docs/src/tutorials/tutorial_morphing.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+Background(1:nframes,(args...)->background("black"))
+circobj = Object((v,o,f) -> circdraw("red"))
+boxobj  = Object((v,o,f) -> boxdraw("green"))
+
+transform_to_box = Action(20:nframes-20, morph_to(boxobj))
+act!(circobj, transform_to_box)
+render(video,pathname="circ_to_box.gif")
+```
+
+![](../assets/circ_to_box.gif)
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:47]: here the gif seems to be the correct one not as in the PR description.

File: docs/src/tutorials/tutorial_morphing.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+if you aren't familiar with this syntax `(v,o,f)-> circdraw("red")` its an "anonymous" function or sometimes called a lambda function.
+Basically a nameless function that is written on the spot in that line of code . One might aswell use any other function `func` in place of it
+(which takes atleast 3 arguments `v,o,f`). Elsewhere in the docs/tutorials you will come across
+something of the form `Object( (args...) -> (some;code;here) )`. This is [splatting](https://docs.julialang.org/en/v1/manual/faq/#The-two-uses-of-the-...-operator:-slurping-and-splatting) and is similar to packing `*args` in python.
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:52]: the semicolons are confusing as it has nothing to do with keyword arguments.

File: docs/src/tutorials/tutorial_morphing.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+Like other animations `morph_to(::Object)` is to be used with action. To learn more about Actions refer to [Tutorial 5](tutorial_5.md)
+Here is a simple code snippet on how to use `morph_to`
+```julia
+using Javis
+
+video = Video(500,500)
+nframes = 160 
+
+function circdraw(colo)
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:22]: Is there a reason why you use `colo` and not `color`?

File: docs/src/tutorials/tutorial_morphing.md [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+act!(circobj, transform_to_box)
+#act!(boxobj, hide_action)
+
+render(video,pathname="circ_to_box_hidden.gif")
+```
+
+![](../assets/circ_to_box_hidden.gif)
+
+Here we have morphed  the circle  without defining an object to morph to. Rather the shape it has to morph into
+is given by a `Function`.
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:103]: we should have another heading here to jump to.

File: docs/src/tutorials/tutorial_morphing.md [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    star(O,100,5,0.5,0.0,:stroke)
+end
+
+Background(1:nframes+10,(args...)->background("black"))
+boxobj = Object(1:nframes+10 , (args...) -> boxdraw("green") )
+
+anim = Animation([0,1],[(boxdraw,["green"]),(circdraw,["red"])])
+
+action = Action(1:nframes,anim,morph())
+act!(boxobj,action1)
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:147]: there is no `action1`

**Review by Wikunia — COMMENTED** [review]
Some additional comments

File: src/morphs.jl (hunk 9 above)
Comment (ArbitRandomUser) [reply src/morphs.jl:106]: > It's very confusing to have this function in the middle of another function. It doesn't seem to work like this for your first example as well. Shouldn't it be `object.opts[:original_func](args...)`?

yes i think it should `ret = object.opts[:original_func](args...)` 

i thought it would be more clear than having an anonymous function . 
Should i change it to ...
```
object.func = (args...) -> begin
drawjpaths...
...
...
end 
```


> Additionally the code you have in the PR description will not produce the gif there, right? Or why should the green square not be there the whole time?

Yes , uploaded the wrong gif , ill fix that

File: docs/src/tutorials/tutorial_morphing.md (hunk 13 above)
Comment (ArbitRandomUser) [reply docs/src/tutorials/tutorial_morphing.md:22]: its kind of a force of habit ,
Although im not using `Colors` , when i do it to avoid name clashes since 
Colors brings in the `color` function into the namespace . 
Its safe to use colors here though

File: docs/src/tutorials/tutorial_morphing.md (hunk 10 above)
Comment (ArbitRandomUser) [reply docs/src/tutorials/tutorial_morphing.md:7]: was not sure how to write a method which doesn't overwrite the existing method for the old morph

**Comment by Wikunia** [comment]
@ArbitRandomUser please let us know whenever you want to have another review of this PR. With a small comment here with things you have changed/added.

File: docs/src/tutorials/tutorial_morphing.md [hunk 16]
```diff
@@ -0,0 +1,194 @@
+# **Tutorial 9:** Morphing `Javis` Objects   
+<!-- TODO test all examples once more and update if they dont work -->
```
Comment (Wikunia) [inline docs/src/tutorials/tutorial_morphing.md:2]: Maybe do this first before trying to shift them in a way to have a better morphing experience. Would like to merge this branch first to keep it a bit smaller.

File: src/luxor_overrides.jl [hunk 17]
```diff
@@ -280,3 +281,254 @@ function background(background_color)
...
+end
+
+function Luxor.fillpreserve()
+    update_currentjpath(:fill)
+    if !DISABLE_LUXOR_DRAW
+        Luxor.Cairo.fill_preserve(Luxor.get_current_cr())
+    end
+end
+
+function Luxor.pathtopoly(::Val{:costate})
```
Comment (Wikunia) [inline src/luxor_overrides.jl:385]: can you add a docstring for this function? Probably also for the ones above but this one isn't that obvious to me just from the function name and argument.

File: src/luxor_overrides.jl [hunk 18]
```diff
@@ -280,3 +281,254 @@ function background(background_color)
...
+    @assert length(loop1) == length(loop2)
+    result = Point[]
+    eased_k = easingfunction(k, 0.0, 1.0, 1.0)
+    for j in 1:length(loop1)
+        push!(result, between(loop1[j], loop2[j], eased_k))
+    end
+    return result
+end
+#this is lifted from luxor, we should ask cormullion if its okay
+function polymorph_noresample(
```
Comment (Wikunia) [inline src/luxor_overrides.jl:465]: Also a docstring 😄

**Review by Wikunia — COMMENTED** [review]
Some minor things. Hope I can test it out over the weekend.

**Comment by ArbitRandomUser** [comment]
@Wikunia @gpucce @Sov-trotter @TheCedarPrince 
The PR at this state has quite a few features implemented . It might be a good idea to review and merge .

It'd be great if you'll can test it out on corner cases you'll come up with ...
or just describe them here , i'll write code to test them out.

I'll keep going through the code and making fixes as I find them.

File: src/Javis.jl [hunk 19]
```diff
@@ -698,6 +699,7 @@ function set_object_defaults!(object)
     scaleto(desired_scale)
 end
 
+#todo add stroke/fills here
 const LUXOR_DONT_EXPORT = [
     :boundingbox,
     :Boxmaptile,
```
Comment (Wikunia) [inline src/Javis.jl:705]: It shouldn't export:
- `strokepath`
- `strokepreserve`
- `fillpath`
- `fillpreserve`
otherwise compiling is broken as a method gets overwritten

**Review by Wikunia — CHANGES_REQUESTED** [review]
Thanks for the ping. I did a code review and experimented a bit with different morphings.
Besides the findings I tagged directly at the code I found the following:
Do you use some luxor methods that were introduced in a version after 2.12? Then we would need to increase the lower bound in the Project.toml. If you're unsure please test it locally with Luxor 2.12 to see if everything passes.

I'm a bit unsure whether I understood how morphing works:
Each object has a jpath which is stored and when morphing is applied the jpath of both the start and finish object or function are sampled to 100 points. Then they get mapped 1-1 and moved along a linear line between their start and end position?

**Comment by ArbitRandomUser** [comment]
Yes i am using some functions from version above 2.12.
All this while my tests were run on 2.19. Tests fail on 2.12
The particular function ended up using is `get_current_strokescale()` in `strokepath`. Which is also used internally in the newer version of Luxor's `strokepath`. We can revert to 2.12 and i can remove the offending lines but morphs where `strokescale` is changed might not be possible ( strokescale morphing is a TODO , not yet implemented). I would say we bump to a higher version (unless you are aware of other issues) .

Yes , your understanding of morphing is correct.  Although i think you know the difference between a JPath and a poly , and that `JPath` contains an array of poly . In-case this is the source of confusion , I'll just be a little more pedantic to correct you that every `poly` in the `JPath` is re-sampled to 100 points . 

If there are unequal number of `JPath`s in the from_object and the to_object. The extra `JPath`s are morphed to/from `null_jpaths` (`JPath`s which if drawn is basically a really small point).


I'll just explain why i think `JPath` is an appropriate term .
When we call `pathtopoly` , what we get is an "`Array` of polys" i.e `Vector{Vector{Point}` from the current `path`. 
The current `path` can have many `subpaths` each of which get converted to 1 poly and it is in this sense that i named `JPath` as the collection of polys .

**Comment by Wikunia** [comment]
Yes I'm fine with setting the lower bound of Luxor even to 3.0

File: src/Javis.jl (hunk 19 above)
Comment (ArbitRandomUser) [reply src/Javis.jl:705]: if we dont export them, wont users have to always use 
`Luxor.strokepath()` instead of just `strokepath()`?

Comment (Wikunia) [reply src/Javis.jl:705]: You need to add them to the list in line 750/751

Comment (ArbitRandomUser) [reply src/Javis.jl:705]: i get undefvarerror when i do that

**Comment by ArbitRandomUser** [comment]
tests will fail now , since changes are made to Luxor

**Comment by ArbitRandomUser** [comment]
use https://github.com/ArbitRandomUser/Luxor.jl/tree/javis_branch2 for testing

File: src/luxor_overrides.jl [hunk 20]
```diff
@@ -280,3 +283,249 @@ function background(background_color)
...
+Arguments
+loop1: first polygon 
+loop2: second polygon 
+k: interpolation factor
+"""
+function _betweenpoly_noresample(
+    loop1,
+    loop2,
+    k,
+    offset = (:former, 1);
```
Comment (ArbitRandomUser) [inline src/luxor_overrides.jl:383]: TODO add this arg to docstring

File: src/luxor_overrides.jl [hunk 21]
```diff
@@ -280,3 +283,249 @@ function background(background_color)
...
+  By default, easingfunction = easingflat, so the intermediate steps are linear. If you use
+  another easing function, intermediate steps are determined by the value of the easing
+  function at k.
+
+"""
+function polymorph_noresample(
+    pgon1::Array{Array{Point,1}},
+    pgon2::Array{Array{Point,1}},
+    k,
+    offsets = Array{Int};
```
Comment (ArbitRandomUser) [inline src/luxor_overrides.jl:442]: TODO add this arg to dosctring

File: src/luxor_utils.jl [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+`action` is either of the four `:fill,:stroke,:fillpreserve,:strokepreserve`.
+and it executes the respective Luxor action . Behaviour depends on two globals
+if CURRENT_FETCHPATH_STATE is true it converts the current path to a JPath and 
+appends to CURRENT_JPATH. 
+if DISABLE_LUXOR_DRAW is true , does not draw anything on to the canvas.
+
+This function is called from Luxor (check luxors source of `strokepath`).
+"""
+function javis_do_action(action::Symbol)
```
Comment (ArbitRandomUser) [inline src/luxor_utils.jl:12]: remove this function , since not really used anymore

File: src/morphs.jl [hunk 23]
```diff
@@ -1,52 +1,324 @@
...
+end
+
+function morph_to(to_func::Function, args = []; samples = 100)
+    return (video, object, action, frame) -> begin
+        action.keep = false
+        _morph_to(video, object, action, frame, to_func, args, samples)
+    end
 end
 
+# a jpath to appear from/disappear into
```
Comment (ArbitRandomUser) [inline src/morphs.jl:96]: make this a docstring

File: src/morphs.jl [hunk 24]
```diff
@@ -1,52 +1,324 @@
...
+
+returns an Array of Tuples , each Tuple is of the form `(s::Symbol,offsetvalue::Int)`
+
+while interpolating polys inside the jpath. Javis tries to find a good offsetvalue
+if poly1 is being morphed into poly2 
+`poly1[i]` goes to `poly2[i + offsetvalue -1]` (modulo length(poly2) addition).
+
+`s` is either `:former` or `:latter` indicating if the offset should be applied on poly1 or poly2
+
+morphing from closed to open closed offsets the former.
```
Comment (ArbitRandomUser) [inline src/morphs.jl:187]: morphing from closed to closed

File: src/luxor_overrides.jl [hunk 25]
```diff
@@ -1,3 +1,6 @@
+DISABLE_LUXOR_DRAW = false
+CURRENT_FETCHPATH_STATE = false
+
```
Comment (Wikunia) [inline src/luxor_overrides.jl:3]: Can you add a description of those here?

File: src/luxor_overrides.jl [hunk 26]
```diff
@@ -280,3 +283,258 @@ function background(background_color)
...
+                    easingfunction = easingfunction,
+                ),
+            )
+            centroid2 = polycentroid(pgon2[i])
+        end
+    end
+    return result
+end
+
+struct JavisLuxorDispatcher <: Luxor.LDispatcher end
```
Comment (Wikunia) [inline src/luxor_overrides.jl:522]: Can we put this struct into src/Javis.jl ? Then all of structs are in one place

File: src/luxor_utils.jl (hunk 22 above)
Comment (Wikunia) [reply src/luxor_utils.jl:12]: Can this be removed?

**Review by Wikunia — COMMENTED** [review]
In general it's good to go from my side. The last time I've run it everything worked out very well and I would like to merge this soon to avoid this getting bigger. @cormullion is there a blocker for releasing a new version of Luxor to have this dispatch functionality that @ArbitRandomUser added to Luxor? Please let us know if there is something we can do to speed it up. 
Thanks again for allowing this change to Luxor.

**Comment by cormullion** [comment]
Hi! There's no reason to not have a Luxor release soon - just that the most recent release was 14 days ago so perhaps I was thinking to get the current changes done and then go for 3.5 in the next few days. Currently there's a PR for adding thread-safety and multiple drawings about to be merged, which probably justifies a new release anyway... 😀

The new release of MathTeXEngine (0.5) has broken LaTeX string drawing in Luxor.text() 😬, but it's not a direct dependency so we may just have to cope with the test and doc failures, or add a compat entry somewhere...

**Comment by Wikunia** [comment]
Thanks for the fast response. Perfect like looking forward to thread safety as well 😊

File: src/structs/Object.jl [hunk 27]
```diff
@@ -110,6 +114,41 @@ function Object(frames, func::Function, start_pos::Union{Object,Point}; kwargs..
...
+This is called at rendertime . `getjpaths!` updates `obj.jpaths` with jpaths that
+are generated by drawing `func` . Typically `func` is `obj.opts[:original_func]`
+"""
+function getjpaths!(obj::Object, func::Function, args = [])
+    m = getmatrix()
+    setmatrix([1.0, 0, 0, 1.0, 0, 0])
+    empty!(CURRENT_JPATHS)
+    global CURRENT_FETCHPATH_STATE = true
+    global DISABLE_LUXOR_DRAW = true
+    v, o, f = nothing, nothing, nothing
```
Comment (gpucce) [inline src/structs/Object.jl:129]: Hi, can't you pass `v, o, f` explicitly as arguments to `getjpaths` instead of defining them as `nothing, nothing, nothing`? It looks they are always available where `getjpaths!` is called

**Comment by Wikunia** [comment]
I think we can now set Luxor to 3.5 and run the tests 😄

**Comment by ArbitRandomUser** [comment]
> I think we can now set Luxor to 3.5 and run the tests smile

tests failed ,idk much about CI but it looks like for some reason the CI seems to not find 3.5 ?

**Comment by Wikunia** [comment]
Yeah Luxor.jl needs Julia 1.6 now which is fine as it's the LTS

**Comment by ArbitRandomUser** [comment]
@Wikunia that didnt work .

**Comment by Wikunia** [comment]
You need to change it here as well: https://github.com/JuliaAnimators/Javis.jl/blob/main/.github/workflows/ci.yml#L12

File: CHANGELOG.md [hunk 28]
```diff
@@ -2,7 +2,14 @@
 
 # PR changes
 - changed render method for mp4 to use ffmpeg directly inplace of VideoIO
-
+- Added jpaths a field in Object that is usefull for morphs and partial drawing
```
Comment (Wikunia) [inline CHANGELOG.md:5]: *useful

File: src/luxor_overrides.jl [hunk 29]
```diff
@@ -280,3 +296,256 @@ function background(background_color)
...
+    @assert length(polygonlist) == length(co_states)
+    #"""return polygonlist, and its closed/open state"""
+    return polygonlist, co_states
+end
+
+
+"""
+    _betweenpoly_noresample(loop1,loop2,k; easingfunction = easingflat)
+
+Just like _betweenpoly from Luxor , but expects polygons `loop1` and `oop2` to be of same size , and
```
Comment (Wikunia) [inline src/luxor_overrides.jl:380]: *loop2

File: src/partial_draw.jl [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+    end
+    #isapprox(fraction, 1) && (object.func = object.opts[:original_func])
+end
+
+"""
+    len_jpath(jpath::JPaht)
+
+returns the sum of all the lengths of the polys that this `jpath` contains
+"""
+len_jpath(jpath::JPath) = sum([
```
Comment (Wikunia) [inline src/partial_draw.jl:94]: no need to create an array here

File: src/structs/JPath.jl [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+    #TODO transform and dashstyle
+    #maybe dont have to store the transform here, just apply the transform
+    #on the polys before storing it ?. This way things like bounding
+    #boxes of the object can be computed easily at compute time.
+end
+
+JPath(polys, closed, fill, stroke, lastaction, linewidth) =
+    JPath(polys, closed, fill, stroke, lastaction, linewidth, nothing)
+
+CURRENT_JPATHS = JPath[] #TODO change to const later
```
Comment (Wikunia) [inline src/structs/JPath.jl:43]: can this be changed to `const` ?

File: src/structs/Object.jl [hunk 32]
```diff
@@ -1,3 +1,4 @@
+include("JPath.jl")
```
Comment (Wikunia) [inline src/structs/Object.jl:1]: let's keep all includes in one place, similar to the structs

File: test/morphing_old.jl [hunk 33]
```diff
@@ -0,0 +1,217 @@
+astar(args...; do_action = :stroke) = star(Point(-100, -100), 30, 5, 0.5, 0, do_action)
```
Comment (Wikunia) [inline test/morphing_old.jl:1]: can this file be removed seems like it's not used, right?

**Review by Wikunia — APPROVED** [review]
Just some minor things but afterwards I would like to merge. 
Thanks a lot for your awesome work. Looking forward to use this functionality.

**Comment by TheCedarPrince** [comment]
Hey! Just did a first pass review of the PR - WOW.
@ArbitRandomUser and @Wikunia , this is absolutely incredible!
If it is alright, may I have one more day to do a few edits to the tutorial and try to follow along it?
Was also going to finish a run through of documentation. 
Otherwise, incredible work!
